common/config/rush/pnpm-lock.yaml

```diff
@@ -1320,7 +1488,7 @@ packages:
     dependencies:
       '@rollup/pluginutils': 3.1.0_rollup@1.32.1
       '@types/resolve': 1.17.1
-      builtin-modules: 3.2.0
+      builtin-modules: 3.1.0
```

Comment:
weird to see: another downgraded package

Reply:
Hi @octogonz, could you please explain how the `--full` param works? I expected it to always upgrade dependencies versions given the description in the docs.

Reply:
We have occasionally seen `rush update --full` downgrade packages in the past.  They typically involve `peerDependencies`.  We will need to investigate each of these downgrades and try to figure out why they might be happening.

Reply:
I think the issue with the package downgrades is our `pnpm-lock.yaml` contains multiple versions of the same package which match the requested version spec, and this seems to trigger non-deterministic behavior in rush+pnpm.

For `builtin-modules`, our `pnpm-lock.yaml` contains both `3.1.0` and `3.2.0`.  `3.2.0` is the latest, but `dev-tool` pins to `~3.1.0`:

https://github.com/Azure/azure-sdk-for-js/blob/5a7c12997212feaf228509063cde054bf318ed4a/common/tools/dev-tool/package.json#L64

If we changed `dev-tool` to depend on `^3.1.0` I believe this problem would go away.

Reply:
I'm still not sure why rush+pnpm isn't always choosing the highest-version matching dependency though.  Might be a bug in pnpm, and might be fixed in v6 (we are still using v5).

Reply:
`rush update --full` effectively deletes `pnpm-lock.yaml` so that when PNPM is choosing versions, it will start from scratch and ignore any previous decisions from the old lockfile.  As you might expect, mostly this causes versions to get upgraded.  But not always, because choosing versions is a complex heuristic where the package manager has a lot of descretion.

As one example, suppose that:

- the latest version of `A` is `1.2.3` 
- `B` wants `1.2.2`
- `C` wants `^1.0.0`
- `D` depends on `B` and `C`

If we always choose the latest version, then we would get side-by-side installations of `1.2.2` and `1.2.3`.  Whereas if we chose `1.2.2` then both `B` and `C` can share the same installation of `A`.  This choice installs a bit faster, and it also avoids problems like singletons from `A` having two instances, or TypeScript typings from `1.2.3` being incompatible with `1.2.2`.  More importantly, for the NPM/YarnClassic `node_modules` plan (i.e. without PNPM's symlinks) insisting on the latest version would unavoidably require an explosion of [dopplegangers](https://rushjs.io/pages/advanced/npm_doppelgangers/).

In short, NPM and SemVer define `^1.0.0` to mean `a version >=1.0.0 and <2`, not necessarily "the latest version."  This gives package managers a lot of leeway to selectively downgrade versions in cases where it would simplify the installation.

Reply:
@octogonz: Thanks for the explanation, that all makes sense.  The issue we seem to be having is that `rush update --recheck` and `rush update --full` sometimes choose different versions, which causes our `pnpm-lock.yaml` to ping-pong between the versions depending which command was run last.

I tried to create a small repro app but was unable to find the right combination of packages to trigger this behavior.  I'm not sure if we can repro reliably in our full monorepo either.  If we can repro this behavior in our full monorepro would you be interested in investigating?

Reply:
I found a repro using our monorepo and opened an issue: https://github.com/microsoft/rushstack/issues/2876

Reply:
Thanks, I'll reply in your issue. 👍